torch_geometric/loader/link_neighbor_loader.py

```diff
@@ -52,43 +54,88 @@ def _create_label(self, edge_label_index, edge_label):
 
         return edge_label_index, edge_label
 
-    def __call__(self, query: List[Tuple[Tensor]]):
-        query = [torch.tensor(s) for s in zip(*query)]
-        if len(query) == 2:
-            edge_label_index = torch.stack(query, dim=0)
-            edge_label = None
+    def _get_data_query(self, edge_label_index):
+        query_nodes = edge_label_index.view(-1)
+        query_nodes, reverse = query_nodes.unique(return_inverse=True)
+        edge_label_index = reverse.view(2, -1)
+        return query_nodes, edge_label_index
+
+    def _get_hetero_data_query(self, edge_label_index):
+
+        if self.input_type[0] != self.input_type[-1]:
+            query_src = edge_label_index[0]
+            query_src, reverse_src = query_src.unique(return_inverse=True)
+            query_dst = edge_label_index[1]
+            query_dst, reverse_dst = query_dst.unique(return_inverse=True)
+            edge_label_index = torch.stack([reverse_src, reverse_dst], 0)
+            query_node_dict = {
+                self.input_type[0]: query_src,
+                self.input_type[-1]: query_dst,
+            }
+        else:  # Merge both source and destination node indices:
+            query_nodes = edge_label_index.view(-1)
+            query_nodes, reverse = query_nodes.unique(return_inverse=True)
+            edge_label_index = reverse.view(2, -1)
+            query_node_dict = {self.input_type[0]: query_nodes}
+        return query_node_dict, edge_label_index
+
+
+class LinkSampler(LinkSamplerMixin):
+    """This sampler is user when :obj:`data` has no
+    :obj:`edge_index`."""
+    def __init__(self, data, neg_sampling_ratio: float = 0.0,
+                 input_type: InputEdges = None):
+        super().__init__()
+        self.data_cls = data.__class__ if isinstance(
+            data, (Data, HeteroData)) else 'custom'
+        self.neg_sampling_ratio = neg_sampling_ratio
```

Comment:
is this used in the `LinkSampler`?

Reply:
Do you mean the `neg_samling_ratio` or `data_cls`?
`data_cls` is used in the `__call__`.